### ISSUE TITLE
Provided chainId "x" must match the active chainId "x"

### ISSUE THREAD

**Issue by joehawary** [MetaMask/metamask-extension#11296]
**Describe the bug**
When performing a Signed transaction using signTypedData an error is thrown.

This includes when passing ChainID as an INT, a HEX, a BigNumber.

This issue has been introduced in 9.6.0 and was not present in 9.5.9.

**Steps to reproduce (REQUIRED)**
1. perform _signTypedData with domain including ChainId
2. Returned error Provided chainId "x" must match the active chainId "x"

**Expected behavior**
No error

**Screenshots**
![Screenshot 2021-06-15 at 14 15 22](https://user-images.githubusercontent.com/32411671/122059307-65421800-cde4-11eb-80c5-a3d99b801629.png)

![Screenshot 2021-06-15 at 10 14 36](https://user-images.githubusercontent.com/32411671/122026821-80039500-cdc2-11eb-97be-e048aad6a502.png)

![Screenshot 2021-06-15 at 10 18 23](https://user-images.githubusercontent.com/32411671/122027473-0d46e980-cdc3-11eb-8985-dd066f40455a.png)

**Browser details (please complete the following information):**
 - OS: Mac
 - Hardware Wallet: None
 - Browser: Edge Chromium & Chrome
 - MetaMask Version: 9.6.0

**Comment by joehawary** [comment]
I've also tried as a string "4"

Even passing the hex as a string "0x4" (for chainID 4)

**Comment by joehawary** [comment]
Also getting the ChainId from the signer and passing it back is the same.

**Comment by joehawary** [comment]
Trials of data being passed:


BigNumber/Hex
![Screenshot 2021-06-15 at 10 43 08](https://user-images.githubusercontent.com/32411671/122031328-81cf5780-cdc6-11eb-8463-164e5579a9bc.png)

INT
![Screenshot 2021-06-15 at 10 42 41](https://user-images.githubusercontent.com/32411671/122031336-83991b00-cdc6-11eb-9bd3-533b0ca7ae90.png)

String
![Screenshot 2021-06-15 at 10 43 59](https://user-images.githubusercontent.com/32411671/122031422-9ad80880-cdc6-11eb-9dac-a03482a977ff.png)

**Comment by joehawary** [comment]
Okay, so after further investigation, it only works if you remove the ChainId entirely from the signing request.

I believe this will break compatibility to previous versions.

**Comment by joehawary** [comment]
^^ Okay further from my last post, when removing the ChainId from the Signing requests, this ends up with an INVALID_SIGNATURE issue. So I would say that removing the ChainId is not an option

**Comment by PierreJeanjacquot** [comment]
same issue here

**Comment by joehawary** [comment]
I've pinned it down to the assertion due to how JSON data is parsed.

https://github.com/MetaMask/metamask-extension/blob/df6c6a46d76da2885c3235961f78794751592e55/app/scripts/lib/typed-message-manager.js#L206-L209

We're just working on a getting a clone of the repo up for debugging

**Comment by adonesky1** [comment]
@joehawary and @PierreJeanjacquot are you using  `_signTypedData` or `_signTypedDataV4`, if the former have you tried the latter yet?

**Comment by adonesky1** [comment]
Actually you must be using V3 or V4 if you're hitting the error you're hitting. Can you pass along the full message params you are using when you get this error?

**Comment by joehawary** [comment]
We are using ```_signTypedData```

Here is an example for you
Domain
```{"name":"Uniswap V2","version":"1","chainId":4,"verifyingContract":"0xE55687682FdF08265d1672EA0C91fA884CcD8955"}```

Types
```{"Permit":[{"name":"owner","type":"address"},{"name":"spender","type":"address"},{"name":"value","type":"uint256"},{"name":"nonce","type":"uint256"},{"name":"deadline","type":"uint256"}]}```

Values
```{"owner":"0x609a29a77c86A1c98a4a907b2C845A02dCfd0727","spender":"0x0ADc14De42436aD95747a1C9A8002D4E60888ACa","value":{"type":"BigNumber","hex":"0x038d7ea4c68000"},"nonce":{"type":"BigNumber","hex":"0x53"},"deadline":1623856041}```

**Comment by adonesky1** [comment]
Awesome thanks. Would you mind sending along just an example of how you are calling the method as well?

**Comment by joehawary** [comment]
```export const signPermitEIP2612 = async (
  owner: Signer,
  walletAddress: string,
  token: Contract,
  spenderAddress: string,
  value: BigNumberish,
  deadline: BigNumberish,
  nonce?: BigNumberish
) => {
  // get nonce
  nonce = nonce || (await token.nonces(walletAddress));
  // get chainId
  const chainId = (await token.provider.getNetwork()).chainId;
  // get domain
  const domain = {
    name: 'Uniswap V2',
    version: '1',
    chainId: chainId,
    verifyingContract: token.address,
  };
  // get types
  const types = {} as Record<string, TypedDataField[]>;
  types['Permit'] = [
    { name: 'owner', type: 'address' },
    { name: 'spender', type: 'address' },
    { name: 'value', type: 'uint256' },
    { name: 'nonce', type: 'uint256' },
    { name: 'deadline', type: 'uint256' },
  ];
  // get values
  const values = {
    owner: walletAddress,
    spender: spenderAddress,
    value: value,
    nonce: nonce,
    deadline: deadline,
  };

  console.log(JSON.stringify(domain))
  console.log(JSON.stringify(types))
  console.log(JSON.stringify(values))

  // sign permission
  // todo: add fallback if wallet does not support eip 712 rpc
  const signedPermission = await (owner as any)._signTypedData(
    domain,
    types,
    values
  );
  // split signature
  const sig = splitSignature(signedPermission);
  // return
  return [
    values.owner,
    values.spender,
    values.value,
    values.deadline,
    sig.v,
    sig.r,
    sig.s,
  ];
};
```

**Comment by joehawary** [comment]
This is another example

```export const signPermission = async (
  method: string,
  vault: Contract,
  owner: Signer,
  delegateAddress: string,
  tokenAddress: string,
  amount: BigNumberish,
  vaultNonce: BigNumberish,
  chainId?: BigNumberish
) => {
  // get chainId
  chainId = chainId || (await vault.provider.getNetwork()).chainId;
  // craft permission
  const domain = {
    name: 'UniversalVault',
    version: '1.0.0',
    chainId: chainId,
    verifyingContract: vault.address,
  };
  const types = {} as Record<string, TypedDataField[]>;
  types[method] = [
    { name: 'delegate', type: 'address' },
    { name: 'token', type: 'address' },
    { name: 'amount', type: 'uint256' },
    { name: 'nonce', type: 'uint256' },
  ];
  const value = {
    delegate: delegateAddress,
    token: tokenAddress,
    amount: amount,
    nonce: vaultNonce,
  };
  // sign permission
  // todo: add fallback if wallet does not support eip 712 rpc
  const signedPermission = await (owner as any)._signTypedData(
    domain,
    types,
    value
  );
  // return
  return signedPermission;
};```

**Comment by joehawary** [comment]
@adonesky1  Strangely enough, if you omit the chainId then it will let you sign the transaction. 

This somehow bypasses the assertion

**Comment by PierreJeanjacquot** [comment]
> @joehawary and @PierreJeanjacquot are you using `_signTypedData` or `_signTypedDataV4`, if the former have you tried the latter yet?

I'm using `v4`

**Comment by adonesky1** [comment]
Ok thank you I will keep digging into this!

**Comment by adonesky1** [comment]
I'm really having trouble reproducing this issue. Wondering if you've had a chance to pull down the extension repo and try to debug against your call? If so could you replace line 207 from `chainId` to `Number(chainId)` and see if you still get the same error?

**Comment by joehawary** [comment]
Yes it works when I set it as ```parseInt(chainId, 16)```

> Ok thank you I will keep digging into this!

**Comment by adonesky1** [comment]
Ok thanks for confirming @joehawary, I'm going to put in a PR soon to try to improve our handling of this validation.

**Comment by joehawary** [comment]
@adonesky1 
Do you want me to send you a link that you can use to test?

**Comment by adonesky1** [comment]
@joehawary yea that would be great!

**Comment by joehawary** [comment]
I've made a version of our site that bypasses requirements

https://deploy-preview-154--alchemist-wtf.netlify.app/

Connect your wallet
Go to mint tab
Click Mint

You will get the stringified and original object of all 3 items that are passed in `_signTypedData` in the Console

Obvious one here but make sure you're on 9.6 haha

**Comment by joehawary** [comment]
Just so you know what's being processed

Here is the exact code from the PR for the site you're viewing

https://github.com/alchemistcoin/crucible-frontend/blob/0cb0079913f05c91aadddcfd9cfa97ed5caa94cd/src/contracts/utils.ts#L86-L92

**Comment by adonesky1** [comment]
@joehawary got a hot fix that we should be pushing out soon!

**Comment by joehawary** [comment]
@adonesky1 
Just been reading through the PR & my comment is that the chainId in our case is being passed as a number.

Is there something that we're missing? Have you managed to try the deploy preview and see the console log?

**Comment by PierreJeanjacquot** [comment]
@adonesky1 your patch looks good to me.

I'm using `ethers` to wrap signTypedData, `ethers` always formats `chainId` as a string regardless of the initial format.

Here is a live example showing the issue of `9.6.0`:
https://codesandbox.io/s/signtypeddata-lmbv4

**Comment by jknight** [comment]
I can reproduce this issue in 9.6.0. I retested in 9.6.1 and it works - thanks for the hotfix !

**Comment by undfined** [comment]
> @joehawary got a hot fix that we should be pushing out soon!

Reeeeally hoping to test this today as EIP-712 sigs are critical to our product design :) I see you cut a new release for 9.6.1 is going live today?

**Comment by rrecuero** [comment]
Any update on this?

**Comment by brad-decker** [comment]
9.6.1 includes the fix for this, and you'll start seeing it land in your various browsers imminently. Thank you for your patience!

**Comment by Danhkaka** [comment]
Good

**Comment by danjm** [comment]
closed with v9.6.1
